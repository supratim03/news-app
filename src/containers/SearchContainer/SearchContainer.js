import React, { useEffect } from "react";
import { useDispatch, useSelector } from "react-redux";
import { useNavigate, useParams } from "react-router-dom";
import HeaderComponent from "../../components/HeaderComponent/HeaderComponent";
import LoaderComponent from "../../components/LoaderComponent/LoaderComponent";
import SelectComponent from "../../components/SelectComponent/SelectComponent";
import { getSearchResults, setIsLoading } from "../../store/actions/actions";
import { Card, CardBlock, Container, ContainerHeaderBlock, GridBlock, GridContainer, SelectBlock } from "./SearchContainerStyled";
import '../../styles/HomeContainer.css'
import FooterComponent from "../../components/FooterComponent/FooterComponent";
import { sortDropdownValues } from "../../constants/constant";

const SearchContainer = () => {
    const { searchVal } = useParams();
    const navigate = useNavigate()
    const dispatch = useDispatch();
    const searchResults = useSelector(state =>  state.news.searchResults);
    const isLoading = useSelector(state => state.news.isLoading)

    useEffect(() => {
        dispatch(setIsLoading(true))
        dispatch(getSearchResults(searchVal, 10, 'newest'))
    }, [])

    const handleArticleById = (id, url) => {
        sessionStorage.setItem('apiUrl', url);
        let value = id.replaceAll('/', '-')
        navigate(`/article/${value}`)
    }

    const prepareSearchCards = searchResults.map(card => {
        return (
            <Card 
                key={card.id}
                style={card.fields?.thumbnail == null ? { backgroundColor: '#2455a1' } : { backgroundImage: "url(" + card.fields?.thumbnail + ")" }} 
                className="search-card" 
                onClick={(e) => handleArticleById(card.id, card.apiUrl)}
            >
                <div className="no-image">
                    <div className={card.fields?.thumbnail == null ? "blank-img" : ""}></div>
                </div>
                <div className="card-desc">
                    <div style={{ padding: '10px' }}>
                        <span className="title">
                            {card.webTitle}
                        </span>
                    </div>
                </div>
            </Card>
        )
    })

    const handleFilterRecords = value => {
        dispatch(getSearchResults(searchVal, 10, value))
    }

    return (
        <>
            <HeaderComponent />
            <Container>
                <GridBlock>
                    <ContainerHeaderBlock>
                        Search Results
                    </ContainerHeaderBlock>
                    <SelectBlock>
                        <SelectComponent optionsList={sortDropdownValues} handleFilterRecords={handleFilterRecords}/>
                    </SelectBlock>
                </GridBlock>
            </Container>
            <CardBlock>
                {!isLoading ? <GridContainer>
                    {prepareSearchCards}
                </GridContainer> : <LoaderComponent />}
            </CardBlock>
            {!isLoading && <FooterComponent />}
        </>
    )
}

export default SearchContainer;