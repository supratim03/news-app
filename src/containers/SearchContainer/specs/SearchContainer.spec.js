import React from "react";
import { mount } from "enzyme/build";
import configureStore from "redux-mock-store";
import createSagaMiddleware from "redux-saga";
import { MemoryRouter } from "react-router-dom";
import { Provider } from "react-redux";
import SearchContainer from "../SearchContainer";
import { Card } from "../SearchContainerStyled";

// mock useNavigate
const mockedUsedNavigate = jest.fn();

jest.mock("react-router-dom", () => ({
    ...jest.requireActual("react-router-dom"),
   useNavigate: () => mockedUsedNavigate
}));

describe("Search container", () => {
    const sagaMiddleWare = createSagaMiddleware();
    const mockStore = configureStore([sagaMiddleWare]);

    const initialState = {
        "news": {
            "searchResults": [],
            "isLoading": true
        }
    };
    const store = mockStore(initialState);
    let wrapper;

    beforeEach(() => {
        wrapper = mount(
            <Provider store={store}>
                <MemoryRouter>
                    <SearchContainer />
                </MemoryRouter>
            </Provider>
        );
    });

    it("render correctly", () => {
        expect(wrapper).toMatchSnapshot();
    });
})

describe("Search container with value", () => {
    const sagaMiddleWare = createSagaMiddleware();
    const mockStore = configureStore([sagaMiddleWare]);

    const initialState = {
        "news": {
            "searchResults": [
                {
                  "id": "sport/2023/feb/12/gimmicky-ollie-robinson-grumbles-pink-cricket-ball-cyclone-gabrielle-england-new-zealand",
                  "type": "article",
                  "sectionId": "sport",
                  "sectionName": "Sport",
                  "webPublicationDate": "2023-02-12T09:09:25Z",
                  "webTitle": "‘Gimmicky’: Robinson grumbles at pink ball as cyclone rumbles in New Zealand",
                  "webUrl": "https://www.theguardian.com/sport/2023/feb/12/gimmicky-ollie-robinson-grumbles-pink-cricket-ball-cyclone-gabrielle-england-new-zealand",
                  "apiUrl": "https://content.guardianapis.com/sport/2023/feb/12/gimmicky-ollie-robinson-grumbles-pink-cricket-ball-cyclone-gabrielle-england-new-zealand",
                  "fields": {
                    "thumbnail": "https://media.guim.co.uk/2d388cedece73b9be32ffa6061466c6a122c7b65/0_81_2045_1226/500.jpg"
                  },
                  "isHosted": false,
                  "pillarId": "pillar/sport",
                  "pillarName": "Sport"
                },
                {
                  "id": "world/2023/feb/12/tear-down-these-walls-or-get-used-to-a-world-of-fear-separation-and-division",
                  "type": "article",
                  "sectionId": "world",
                  "sectionName": "World news",
                  "webPublicationDate": "2023-02-12T09:00:19Z",
                  "webTitle": "Tear down these walls, or get used to a world of fear, separation and division | Simon Tisdall",
                  "webUrl": "https://www.theguardian.com/world/2023/feb/12/tear-down-these-walls-or-get-used-to-a-world-of-fear-separation-and-division",
                  "apiUrl": "https://content.guardianapis.com/world/2023/feb/12/tear-down-these-walls-or-get-used-to-a-world-of-fear-separation-and-division",
                  "fields": {
                    "thumbnail": "https://media.guim.co.uk/0ba5fecff1226de8b2df7895b81d22cc0305e4ff/0_38_4926_2956/500.jpg"
                  },
                  "isHosted": false,
                  "pillarId": "pillar/news",
                  "pillarName": "News"
                }
              ],
            "isLoading": false
        }
    };
    const store = mockStore(initialState);
    let wrapper;

    beforeEach(() => {
        wrapper = mount(
            <Provider store={store}>
                <MemoryRouter>
                    <SearchContainer />
                </MemoryRouter>
            </Provider>
        );
    });

    it("render correctly", () => {
        expect(wrapper).toMatchSnapshot();
    });

    it('should navigate on click of card', () => {
        const element = wrapper.find(Card).at(0);
        element.simulate('click', "world/2023/feb/12/bbc-steve-rosenberg-russia-editor-ukraine-war-piano-gorbachev-eurovision", "https://www.theguardian.com/world/2023/feb/12/bbc-steve-rosenberg-russia-editor-ukraine-war-piano-gorbachev-eurovision");
        setTimeout(() => {
            expect(mockedUsedNavigate).toHaveBeenCalled();
        }, 0)
    })
})