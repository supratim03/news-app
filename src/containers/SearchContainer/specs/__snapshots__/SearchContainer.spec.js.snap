// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search container render correctly 1`] = `ReactWrapper {}`;

exports[`Search container with value render correctly 1`] = `ReactWrapper {}`;
